File: packages/jev-router/policy.mjs
/**
 * Versioned thresholds. Jev classifies; code decides.
 * Pattern source: Stanley (MIT) route-intent / check-task — reimplemented thin, not vendored.
 */

export const ROUTING_POLICY = {
  version: "omapi-route-workflow-policy@1",
  minConfidence: 0.6,
  minProbability: 0.55,
  minMargin: 0.15,
};

export const CHECK_POLICY = {
  version: "omapi-check-policy@1",
  maxHunks: 32,
  maxHunkChars: 4000,
  concernFinding: 0.7,
  concernPark: 0.4,
  kindMinConfidence: 0.6,
  kindMinProbability: 0.55,
};

/**
 * Loop-stop / escalate (Suraj #4 + #10). Sibling of the routing policy.
 * Jev classifies continue|stop|escalate; code applies thresholds.
 * Escalate is HITL — never auto-retry or auto-promote.
 */
export const LOOP_STOP_POLICY = {
  version: "omapi-loop-stop-policy@1",
  minConfidence: 0.6,
  minProbability: 0.55,
  minMargin: 0.15,
  maxDigestChars: 4000,
};

export const WORKFLOW_LABELS = ["check", "review", "cannot_tell"];
export const LOOP_STOP_LABELS = ["continue", "stop", "escalate"];
export const CHECK_KIND_LABELS = ["none", "test_safety", "task_mismatch", "cannot_tell"];

/**
 * JEV_BYPASS kill switch. Same spelling as packages/cursor-agent-jev.nix:
 * only "1" or "true". Anything else (unset, "0", "false", "yes") is not a bypass.
 * The wrap then execs cursor-agent. The Grok hook must defer, not emit allow.
 */
export function isJevBypass(value) {
  return value === "1" || value === "true";
}

/**
 * Shared exec table for a GateVerdict. Shadow always execs.
 * Active: stop/escalate do not; blocked does not; continue or gate=auto does.
 * This is not an auto-allow. Callers that sit in front of a permission prompt
 * must defer, not emit decision "allow".
 */
export function gateVerdictAllowsExec({ mode, choice, gate, blocked } = {}) {
  const m = String(mode || "shadow").toLowerCase();
  if (m !== "active") return true;
  const c = String(choice || "");
  const g = String(gate || "");
  const b = blocked === true || blocked === "true";
  if (c === "stop" || c === "escalate") return false;
  if (b) return false;
  return c === "continue" || g === "auto";
}

/**
 * Map a jev-router GateVerdict onto one Grok PreToolUse decision.
 * Code denies. Bypass wins over a stop-shaped payload (the wrap skips the router).
 * Non-deny is `defer` — never `allow`.
 *
 * `JEV_MODE` shadow defers unless `permission.honor` or `calls.honor` is set.
 * Permission honor is `JEV_PERMISSION_MODE=active` and a key. Calls honor is
 * `JEV_TYPED_CALL_MODE=active` and a key. A continue does not override a loop
 * stop. A stop/escalate on either honoring surface denies even when the loop
 * would exec. The decision is still defer or deny, never allow.
 */
export function pretoolHookDecision(verdict = {}) {
  if (verdict.bypass === true) {
    return { action: "defer", exitCode: 0, decision: "defer", reason: "bypass" };
  }

  const permission = verdict.permission && typeof verdict.permission === "object" ? verdict.permission : null;
  const permissionHonors = Boolean(
    permission && permission.honor === true && String(permission.mode || "").toLowerCase() === "active",
  );
  const calls = verdict.calls && typeof verdict.calls === "object" ? verdict.calls : null;
  const callsHonors = Boolean(calls && calls.honor === true && String(calls.mode || "").toLowerCase() === "active");
  const modeActive = String(verdict.mode || "shadow").toLowerCase() === "active";
  if (!modeActive && !permissionHonors && !callsHonors) {
    return { action: "defer", exitCode: 0, decision: "defer", reason: "shadow" };
  }

  const loopAllows = gateVerdictAllowsExec({ ...verdict, mode: "active" });
  const permBlocks = Boolean(
    permissionHonors &&
      (permission.blocked === true || permission.choice === "stop" || permission.choice === "escalate"),
  );
  const callsBlocks = Boolean(
    callsHonors && (calls.blocked === true || calls.choice === "stop" || calls.choice === "escalate"),
  );

  let choice = verdict.choice != null ? String(verdict.choice) : "unclassified";
  let gate = verdict.gate != null ? String(verdict.gate) : "hold";
  let blocked = verdict.blocked === true || verdict.blocked === "true";
  let allows = false;

  if (permBlocks || callsBlocks) {
    const blocker = permBlocks ? permission : calls;
    choice = blocker.choice != null ? String(blocker.choice) : "stop";
    gate = blocker.gate != null ? String(blocker.gate) : "hold";
    blocked = true;
    allows = false;
  } else if (!loopAllows) {
    allows = false;
  } else {
    if (permissionHonors && permission.choice) choice = String(permission.choice);
    if (permissionHonors && permission.gate) gate = String(permission.gate);
    blocked = false;
    allows = gateVerdictAllowsExec({ mode: "active", choice, gate, blocked });
  }

  if (allows) {
    return { action: "defer", exitCode: 0, decision: "defer", reason: "exec" };
  }

  const policyId =
    (verdict.pretool && verdict.pretool.policyId) ||
    (permBlocks && permission && permission.policyId) ||
    (callsBlocks && calls && calls.policyId) ||
    (permission && permission.policyId) ||
    "";
  const policy = policyId ? ` policy=${policyId}` : "";
  return {
    action: "deny",
    exitCode: 2,
    decision: "deny",
    reason: `jev choice=${choice} gate=${gate}${policy}`,
  };
}

/**
 * Grok Build PreToolUse tool class → existing omapi policyId / Choice family.
 * One map for the Grok hook (~/.grok/hooks/jev-omapi.json) and this router.
 * Not a new policy and not a second TypeSafe client.
 * Permission catalogs wrap these ids (packages/jev-router/permission.mjs).
 *
 * Shell / web / subagent stay on loop-stop (the exec gate).
 * Write stays on check (the diff/hunk gate).
 * MCP stays on route-workflow (the capability gate). Typed Calls rank tools
 * on that same id; they do not add a policy.
 *
 * MCP names are the qualified `server__tool` form Grok puts on the event
 * (e.g. linear__save_issue), not the `use_tool` dispatcher.
 * Shell includes both public spellings: hook alias `run_terminal_command`
 * and the shell tool id `run_terminal_cmd`.
 */
export const MCP_TOOL_PATTERN = "[A-Za-z0-9][A-Za-z0-9_.-]*__[A-Za-z0-9_.-]+";

export const PRETOOL_CLASSES = [
  {
    id: "web",
    policyId: LOOP_STOP_POLICY.version,
    choiceFamily: LOOP_STOP_LABELS,
    tools: ["web_search", "WebSearch", "web_fetch", "WebFetch"],
  },
  {
    id: "subagent",
    policyId: LOOP_STOP_POLICY.version,
    choiceFamily: LOOP_STOP_LABELS,
    tools: ["spawn_subagent", "Task"],
  },
  {
    id: "shell",
    policyId: LOOP_STOP_POLICY.version,
    choiceFamily: LOOP_STOP_LABELS,
    tools: ["Bash", "run_terminal_command", "run_terminal_cmd"],
  },
  {
    id: "write",
    policyId: CHECK_POLICY.version,
    choiceFamily: CHECK_KIND_LABELS,
    tools: ["Write", "Edit", "MultiEdit", "search_replace"],
  },
  {
    id: "mcp",
    policyId: ROUTING_POLICY.version,
    choiceFamily: WORKFLOW_LABELS,
    tools: [],
    pattern: MCP_TOOL_PATTERN,
  },
];

export const PRETOOL_MATCHER = PRETOOL_CLASSES.flatMap((row) => row.tools).concat(MCP_TOOL_PATTERN).join("|");

const MCP_TOOL_RE = new RegExp(`^${MCP_TOOL_PATTERN}$`);

export function pretoolClassById(id) {
  const key = String(id || "");
  return PRETOOL_CLASSES.find((row) => row.id === key) || null;
}

/** Full-string class match. Null when the tool is outside the PreToolUse map. */
export function matchPretoolClass(toolName) {
  const name = String(toolName || "");
  if (!name) return null;
  for (const row of PRETOOL_CLASSES) {
    if (row.tools.includes(name)) return row;
  }
  if (MCP_TOOL_RE.test(name)) return pretoolClassById("mcp");
  return null;
}

export function pretoolStamp({ toolName = "", toolClass = "" } = {}) {
  const name = String(toolName || "");
  const requested = String(toolClass || "");
  if (!name && !requested) return null;
  const byName = name ? matchPretoolClass(name) : null;
  const byClass = requested ? pretoolClassById(requested) : null;
  const row = byClass || byName;
  const mismatch = Boolean(byClass && byName && byClass.id !== byName.id);
  if (!row) {
    return {
      matched: false,
      class: null,
      policyId: null,
      choiceFamily: [],
      toolName: name || null,
      mismatch,
    };
  }
  return {
    matched: true,
    class: row.id,
    policyId: row.policyId,
    choiceFamily: row.choiceFamily.slice(),
    toolName: name || null,
    mismatch,
  };
}

/**
 * Apply routing thresholds to a Jev Choice answer.
 * Unavailable or uncertain picks become cannot_tell. Code, not the model, decides.
 */
export function applyRoutingThresholds(answer, capabilities, policy = ROUTING_POLICY) {
  const selected = answer && answer.choice != null ? String(answer.choice) : "cannot_tell";
  const probabilities = (answer && answer.probabilities) || {};
  const selectedProb = Number(probabilities[selected] ?? 0);
  const confidence = Number(answer && answer.confidence != null ? answer.confidence : 0);
  const alts = Object.entries(probabilities)
    .filter(([label]) => label !== selected)
    .map(([, value]) => Number(value ?? 0));
  const margin = selectedProb - (alts.length ? Math.max(...alts) : 0);

  const base = {
    policy: policy.version,
    selected,
    confidence,
    selectedProb,
    margin,
  };

  if (selected === "cannot_tell") {
    return { ...base, outcome: "cannot_tell", reason: "cannot_tell" };
  }
  if (capabilities[selected] !== true) {
    return { ...base, outcome: "cannot_tell", reason: "unavailable" };
  }
  if (
    confidence < policy.minConfidence ||
    selectedProb < policy.minProbability ||
    margin < policy.minMargin
  ) {
    return { ...base, outcome: "cannot_tell", reason: "model_uncertain" };
  }
  return { ...base, outcome: selected, reason: "selected" };
}

/**
 * Apply check-policy thresholds to one hunk judgment (noul + kind Choice).
 * Clear hunks are not findings. Empty findings are never approval.
 */
export function applyCheckThresholds(judgment, policy = CHECK_POLICY) {
  const concern = Number(judgment && judgment.concern != null ? judgment.concern : 0);
  const kind = judgment && judgment.kind ? String(judgment.kind) : "cannot_tell";
  const confidence = Number(judgment && judgment.confidence != null ? judgment.confidence : 0);
  const probabilities = (judgment && judgment.probabilities) || {};
  const kindProb = Number(probabilities[kind] ?? 0);

  if (kind === "cannot_tell" || confidence < policy.kindMinConfidence || kindProb < policy.kindMinProbability) {
    return { bucket: "parked", reason: "model_uncertain", flag: null, concern, kind };
  }
  if (concern >= policy.concernFinding) {
    const flag = kind === "none" ? "jev_concern" : kind;
    return { bucket: "finding", reason: "concern_above_threshold", flag, concern, kind };
  }
  if (concern >= policy.concernPark) {
    return { bucket: "parked", reason: "concern_in_band", flag: null, concern, kind };
  }
  return { bucket: "clear", reason: "below_park", flag: null, concern, kind };
}

function choiceStats(answer) {
  const selected = answer && answer.choice != null ? String(answer.choice) : "cannot_tell";
  const probabilities = (answer && answer.probabilities) || {};
  const selectedProb = Number(probabilities[selected] ?? 0);
  const confidence = Number(answer && answer.confidence != null ? answer.confidence : 0);
  const alts = Object.entries(probabilities)
    .filter(([label]) => label !== selected)
    .map(([, value]) => Number(value ?? 0));
  const margin = selectedProb - (alts.length ? Math.max(...alts) : 0);
  return { selected, probabilities, selectedProb, confidence, margin };
}

function meetsChoiceThresholds({ confidence, selectedProb, margin }, policy) {
  return (
    confidence >= policy.minConfidence &&
    selectedProb >= policy.minProbability &&
    margin >= policy.minMargin
  );
}

/**
 * Apply loop-stop thresholds to a Jev Choice over {continue, stop, escalate}.
 * Escalate is sticky: low confidence still HITL, never auto-continue.
 * Uncertain continue/stop become cannot_tell — code, not the model, decides.
 */
export function applyLoopStopThresholds(answer, policy = LOOP_STOP_POLICY) {
  const { selected, selectedProb, confidence, margin } = choiceStats(answer);
  const base = {
    policy: policy.version,
    selected,
    confidence,
    selectedProb,
    margin,
  };
  const confident = meetsChoiceThresholds({ confidence, selectedProb, margin }, policy);

  if (selected === "escalate") {
    return { ...base, outcome: "escalate", reason: confident ? "selected" : "escalate_uncertain" };
  }
  if (selected !== "continue" && selected !== "stop") {
    return { ...base, outcome: "cannot_tell", reason: "cannot_tell" };
  }
  if (!confident) {
    return { ...base, outcome: "cannot_tell", reason: "model_uncertain" };
  }
  return { ...base, outcome: selected, reason: "selected" };
}

File: crates/bezel-bridle/src/policy.rs
//! Thresholds and the hook map. Jev classifies. Code authorizes.
//! No client is constructed here.

pub const CHECK_POLICY_ID: &str = "omapi-check-policy@1";
pub const LOOP_STOP_POLICY_ID: &str = "omapi-loop-stop-policy@1";
pub const ROUTING_POLICY_ID: &str = "omapi-route-workflow-policy@1";
pub const CONCERN_PARK: f64 = 0.4;
pub const CONCERN_FINDING: f64 = 0.7;
pub const KIND_MIN_CONFIDENCE: f64 = 0.6;
pub const KIND_MIN_PROBABILITY: f64 = 0.55;
pub const MAX_HUNK_CHARS: usize = 4000;
pub const MAX_DIGEST_CHARS: usize = 4000;
pub const PRETOOL_MATCHER: &str = "web_search|WebSearch|web_fetch|WebFetch|spawn_subagent|Task|Bash|run_terminal_command|run_terminal_cmd|Write|Edit|MultiEdit|search_replace|[A-Za-z0-9][A-Za-z0-9_.-]*__[A-Za-z0-9_.-]+";
pub const AUTO_ALLOW: bool = false;

pub const LOOP_STOP_MIN_CONFIDENCE: f64 = 0.6;
pub const LOOP_STOP_MIN_PROBABILITY: f64 = 0.55;
pub const LOOP_STOP_MIN_MARGIN: f64 = 0.15;

pub const WRITE_CODE_DENY_FLAGS: &[&str] = &[
    "secret_path",
    "skip_marker_added",
    "assertions_removed",
    "test_file_deleted",
];

/// Active continue defers. This function never returns `allow`.
pub fn hook_decision(mapped: &str, gate: &str) -> &'static str {
    if mapped == "continue" && gate == "auto" {
        "defer"
    } else {
        "deny"
    }
}

pub fn is_jev_bypass(value: &str) -> bool {
    value == "1" || value == "true"
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Stamp {
    pub matched: bool,
    pub class_id: Option<&'static str>,
    pub policy_id: Option<&'static str>,
}

pub fn match_pretool_class(tool_name: &str) -> Option<(&'static str, &'static str)> {
    const NAMED: &[(&str, &str, &str)] = &[
        ("web_search", "web", LOOP_STOP_POLICY_ID),
        ("WebSearch", "web", LOOP_STOP_POLICY_ID),
        ("web_fetch", "web", LOOP_STOP_POLICY_ID),
        ("WebFetch", "web", LOOP_STOP_POLICY_ID),
        ("spawn_subagent", "subagent", LOOP_STOP_POLICY_ID),
        ("Task", "subagent", LOOP_STOP_POLICY_ID),
        ("Bash", "shell", LOOP_STOP_POLICY_ID),
        ("run_terminal_command", "shell", LOOP_STOP_POLICY_ID),
        ("run_terminal_cmd", "shell", LOOP_STOP_POLICY_ID),
        ("Write", "write", CHECK_POLICY_ID),
        ("Edit", "write", CHECK_POLICY_ID),
        ("MultiEdit", "write", CHECK_POLICY_ID),
        ("search_replace", "write", CHECK_POLICY_ID),
    ];
    if let Some((_, class_id, policy_id)) = NAMED.iter().find(|(name, _, _)| *name == tool_name) {
        return Some((*class_id, *policy_id));
    }
    if is_mcp_name(tool_name) {
        return Some(("mcp", ROUTING_POLICY_ID));
    }
    None
}

pub fn pretool_stamp(tool_name: &str) -> Option<Stamp> {
    if tool_name.is_empty() {
        return None;
    }
    match match_pretool_class(tool_name) {
        Some((class_id, policy_id)) => Some(Stamp {
            matched: true,
            class_id: Some(class_id),
            policy_id: Some(policy_id),
        }),
        None => Some(Stamp {
            matched: false,
            class_id: None,
            policy_id: None,
        }),
    }
}

fn is_mcp_name(name: &str) -> bool {
    let Some((left, right)) = name.split_once("__") else {
        return false;
    };
    if left.is_empty() || right.is_empty() {
        return false;
    }
    let mut chars = left.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    if !first.is_ascii_alphanumeric() {
        return false;
    }
    let ok = |c: char| c.is_ascii_alphanumeric() || c == '_' || c == '.' || c == '-';
    left.chars().all(ok) && right.chars().all(ok)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SurfaceVerdict {
    pub honor: bool,
    pub mode: String,
    pub choice: String,
    pub gate: String,
    pub blocked: bool,
    pub policy_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GateVerdict {
    pub bypass: bool,
    pub mode: String,
    pub choice: Option<String>,
    pub gate: Option<String>,
    pub blocked: bool,
    pub policy_id: String,
    pub permission: Option<SurfaceVerdict>,
    pub calls: Option<SurfaceVerdict>,
}

impl GateVerdict {
    pub fn simple(mode: &str, choice: &str, gate: &str, blocked: bool) -> Self {
        Self {
            bypass: false,
            mode: mode.to_string(),
            choice: Some(choice.to_string()),
            gate: Some(gate.to_string()),
            blocked,
            policy_id: String::new(),
            permission: None,
            calls: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HookOut {
    pub decision: &'static str,
    pub reason: String,
    pub exit_code: i32,
}

/// Shadow always execs. Active stop, escalate, and blocked do not.
pub fn should_exec_agent(mode: &str, choice: &str, gate: &str, blocked: bool) -> bool {
    if !mode.eq_ignore_ascii_case("active") {
        return true;
    }
    if choice == "stop" || choice == "escalate" || blocked {
        return false;
    }
    choice == "continue" || gate == "auto"
}

fn allows_exec(choice: &str, gate: &str, blocked: bool) -> bool {
    if choice == "stop" || choice == "escalate" || blocked {
        return false;
    }
    choice == "continue" || gate == "auto"
}

fn surface_honors(surface: &Option<SurfaceVerdict>) -> bool {
    matches!(surface, Some(row) if row.honor && row.mode.eq_ignore_ascii_case("active"))
}

/// Map a gate verdict onto one hook decision. Never `allow`.
pub fn pretool_hook_decision(verdict: &GateVerdict) -> HookOut {
    if verdict.bypass {
        return HookOut {
            decision: "defer",
            reason: "bypass".to_string(),
            exit_code: 0,
        };
    }
    let permission_honors = surface_honors(&verdict.permission);
    let calls_honors = surface_honors(&verdict.calls);
    let mode_active = verdict.mode.eq_ignore_ascii_case("active");
    if !mode_active && !permission_honors && !calls_honors {
        return HookOut {
            decision: "defer",
            reason: "shadow".to_string(),
            exit_code: 0,
        };
    }

    let mut choice = verdict.choice.clone().unwrap_or_else(|| "unclassified".to_string());
    let mut gate = verdict.gate.clone().unwrap_or_else(|| "hold".to_string());
    let loop_allows = allows_exec(&choice, &gate, verdict.blocked);
    let perm_blocks = permission_honors
        && verdict.permission.as_ref().is_some_and(|row| {
            row.blocked || row.choice == "stop" || row.choice == "escalate"
        });
    let calls_blocks = calls_honors
        && verdict.calls.as_ref().is_some_and(|row| row.blocked || row.choice == "stop" || row.choice == "escalate");

if perm_blocks || calls_blocks {
        let blocker = if perm_blocks {
            verdict.permission.as_ref()
        } else {
            verdict.calls.as_ref()
        };
        if let Some(row) = blocker {
            choice = if row.choice.is_empty() {
                "stop".to_string()
            } else {
                row.choice.clone()
            };
            gate = if row.gate.is_empty() {
                "hold".to_string()
            } else {
                row.gate.clone()
            };
        }
    } else if loop_allows && permission_honors {
        if let Some(row) = verdict.permission.as_ref() {
            if !row.choice.is_empty() {
                choice = row.choice.clone();
            }
            if !row.gate.is_empty() {
                gate = row.gate.clone();
            }
        }
    }
    let allows = !perm_blocks && !calls_blocks && loop_allows && allows_exec(&choice, &gate, false);

    if allows {
        return HookOut {
            decision: "defer",
            reason: "exec".to_string(),
            exit_code: 0,
        };
    }

    let policy_id = if !verdict.policy_id.is_empty() {
        verdict.policy_id.clone()
    } else if perm_blocks {
        verdict.permission.as_ref().map(|row| row.policy_id.clone()).unwrap_or_default()
    } else if calls_blocks {
        verdict.calls.as_ref().map(|row| row.policy_id.clone()).unwrap_or_default()
    } else {
        verdict.permission.as_ref().map(|row| row.policy_id.clone()).unwrap_or_default()
    };
    let policy = if policy_id.is_empty() {
        String::new()
    } else {
        format!(" policy={policy_id}")
    };
    HookOut {
        decision: "deny",
        reason: format!("jev choice={choice} gate={gate}{policy}"),
        exit_code: 2,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct RouteDecision {
    pub outcome: String,
    pub reason: &'static str,
    pub selected: String,
    pub confidence: f64,
    pub selected_prob: f64,
    pub margin: f64,
}

pub fn apply_routing_thresholds(
    choice: &str,
    confidence: f64,
    probabilities: &[(&str, f64)],
    available: &[&str],
) -> RouteDecision {
    let selected = if choice.is_empty() { "cannot_tell" } else { choice };
    let selected_prob = probabilities
        .iter()
        .find(|(label, _)| *label == selected)
        .map(|(_, value)| *value)
        .unwrap_or(0.0);
    let runner_up = probabilities
        .iter()
        .filter(|(label, _)| *label != selected)
        .map(|(_, value)| *value)
        .fold(0.0_f64, f64::max);
    let margin = selected_prob - runner_up;
    let (outcome, reason) = if selected == "cannot_tell" {
        ("cannot_tell", "cannot_tell")
    } else if !available.iter().any(|name| *name == selected) {
        ("cannot_tell", "unavailable")
    } else if confidence < LOOP_STOP_MIN_CONFIDENCE
        || selected_prob < LOOP_STOP_MIN_PROBABILITY
        || margin < LOOP_STOP_MIN_MARGIN
    {
        ("cannot_tell", "model_uncertain")
    } else {
        (selected, "selected")
    };
    RouteDecision {
        outcome: outcome.to_string(),
        reason,
        selected: selected.to_string(),
        confidence,
        selected_prob,
        margin,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct CheckBucket {
    pub bucket: &'static str,
    pub reason: &'static str,
    pub flag: Option<String>,
}

pub fn apply_check_thresholds(
    concern: f64,
    kind: &str,
    confidence: f64,
    probabilities: &[(&str, f64)],
) -> CheckBucket {
    let kind = if kind.is_empty() { "cannot_tell" } else { kind };
    let kind_prob = probabilities
        .iter()
        .find(|(label, _)| *label == kind)
        .map(|(_, value)| *value)
        .unwrap_or(0.0);
    if kind == "cannot_tell" || confidence < KIND_MIN_CONFIDENCE || kind_prob < KIND_MIN_PROBABILITY {
        return CheckBucket {
            bucket: "parked",
            reason: "model_uncertain",
            flag: None,
        };
    }
    if concern >= CONCERN_FINDING {
        let flag = if kind == "none" { "jev_concern" } else { kind };
        return CheckBucket {
            bucket: "finding",
            reason: "concern_above_threshold",
            flag: Some(flag.to_string()),
        };
    }
    if concern >= CONCERN_PARK {
        return CheckBucket {
            bucket: "parked",
            reason: "concern_in_band",
            flag: None,
        };
    }
    CheckBucket {
        bucket: "clear",
        reason: "below_park",
        flag: None,
    }
}

pub fn choice_family(class_id: &str) -> &'static [&'static str] {
    match class_id {
        "web" | "subagent" | "shell" => &["continue", "stop", "escalate"],
        "write" => &["none", "test_safety", "task_mismatch", "cannot_tell"],
        "mcp" => &["check", "review", "cannot_tell"],
        _ => &[],
    }
}
